core/frontend/src/main/scala/io/udash/Application.scala

```diff
@@ -9,18 +9,18 @@ import scala.reflect.ClassTag
 /**
   * Root application which is used to start single instance of app.
   *
-  * @param routingRegistry [[io.udash.core.RoutingRegistry]] implementation, which will be used to match [[io.udash.core.Url]] to [[io.udash.core.State]]
-  * @param viewPresenterRegistry [[io.udash.core.ViewPresenterRegistry]] implementation, which will be used to match [[io.udash.core.State]] into [[io.udash.core.ViewPresenter]]
-  * @param rootState The instance of [[io.udash.core.State]] which will treated as main state.
-  * @tparam S Should be a sealed trait which extends [[io.udash.core.State]].
+  * @param routingRegistry     [[io.udash.core.RoutingRegistry]] implementation, which will be used to match [[io.udash.core.Url]] to [[io.udash.core.State]]
+  * @param viewFactoryRegistry [[io.udash.core.ViewFactoryRegistry]] implementation, which will be used to match [[io.udash.core.State]] into [[io.udash.core.ViewFactory]]
+  * @param rootState           The instance of [[io.udash.core.State]] which will treated as main state.
+  * @tparam HierarchyRoot Should be a sealed trait which extends [[io.udash.core.State]].
   */
-class Application[S <: State : ClassTag : ImmutableValue](routingRegistry: RoutingRegistry[S],
-                                                          viewPresenterRegistry: ViewPresenterRegistry[S],
-                                                          rootState: S,
-                                                          urlChangeProvider: UrlChangeProvider = WindowUrlChangeProvider) {
+class Application[HierarchyRoot <: State[HierarchyRoot] : ClassTag : ImmutableValue](routingRegistry: RoutingRegistry[HierarchyRoot],
```

Comment:
This class will be heavily extended by library users and rightfully so. Make some of the methods, which guarantee correct behaviour of core udash features final.